var/spack/repos/builtin/packages/trilinos/package.py

```diff
@@ -368,6 +368,13 @@ class Trilinos(CMakePackage, CudaPackage, ROCmPackage):
 
     # ###################### Dependencies ##########################
 
+    # External Kokkos
+    depends_on("kokkos@4.1.00", when="@14.4.0 +kokkos")
+    for a in CudaPackage.cuda_arch_values:
+        arch_str = "+cuda cuda_arch={0}".format(a)
+        kokkos_spec = "kokkos@4.1.00 {0}".format(arch_str)
+        depends_on(kokkos_spec, when="@14.4.0 +kokkos {0}".format(arch_str))
+
```

Comment:
That is interesting.  When Trilinos depends on Kokkos, it should not even have any specs related to the Kokkos architecture.  But with the need to support older versions of Trilinos that build Kokkos internally, I can see why this is necassary.

Reply:
Are you saying it shouldn't be necessary to build turn on `+cuda` for the Spack-built `kokkos@4.1.00` used to build `triinos@14.4.0 +kokkos +cuda` ? Likewise with other GPU targets?

In other words, building `trilinos@14.4.0 +kokkos +cuda ^kokkos~cuda` is expected to work perfectly fine?

Reply:
I think we need to write these kinds of conflicts in the spec currently. We should modify the `CMakePackage`, `ROCMPackage` etc to auto detect these kind of things though

Reply:
I don't think you can concretize `trilinos@14.4.0 +kokkos +cuda ^kokkos~cuda` thought. It should break because it depends on a `kokkos+cuda` as written

Reply:
> I don't think you can concretize `trilinos@14.4.0 +kokkos +cuda ^kokkos~cuda` thought. It should break because it depends on a `kokkos+cuda` as written

I know. But I am trying to understand what @bartlettroscoe was saying above.

Reply:
I am saying that `trilinos@14.4.0` should not even have a `cuda` or `kokkos` variant.  Trilinos will support CUDA if Kokkos supports CUDA or not, period.  And with `trilinos@14.4.0`  and newer versions, there should be no `kokkos` variant because Trilinos will **always** depend on the Spack `kokkos` package (because you can't build anything useful in Trilinos without Kokkos).  Therefore, the Trilinos package (as of version 14.4+) should not even define a  `cuda` or `kokkos` variant.  But I understand the need to support older versions of Trilinos requires to have a `cuda` variant at the very least.  But why did Trilinos (at least for the last 10 years) ever have a `kokkos` variant?  What can you build in Trilinos without Kokkos?

With that said, it is harmless for Trilinos to have the `kokkos` and `cuda` variants as long as you add the needed constraints.

Reply:
Because Trilinos's package inherits from the `CudaPackage` it will have a cuda variant.  We should write this so that if someone specifies `trilinos+cuda` then the parent `kokkos` will automatically configure cuda as well, and vice versa i.e. if you did `trilinos+kokkos ^kokkos+cuda` it needs to concretize to the same things.

Typically we are just specifying the top level specs, so `kokkos` will get its constraints from the `trilinos` package based on what the user specifies